### PR TITLE
integration test and aws improvements

### PR DESCRIPTION
**Pull request by j4k4**
The test package got some improvements to have less code redundancy and provider better testing suites.
In Addition, this PR contains some code to prepare the switch to the AWS SDK V2.

### CODE REVIEW

File: pkg/test/suite/run.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	}
+}
+
+func suiteFindTestCases(t *testing.T, suite TestingSuite) (map[string]testCaseRunner, error) {
+	var err error
+	var regex *regexp.Regexp
+	var testCases = make(map[string]testCaseRunner)
+	var methodFinder = reflect.TypeOf(suite)
+
+	if regex, err = regexp.Compile("^Test"); err != nil {
```
Comment (ajscholl) [inline pkg/test/suite/run.go:52]: No need for a regex when this is just `strings.HasPrefix`

File: pkg/test/suite/testcase_subscriber.go [hunk 2]
```diff
@@ -38,173 +47,144 @@ func (s subscriberTestCase) GetInput() interface{} {
...
-	attrs := mdlsub.CreateMessageAttributes(sub.GetModelId(), "create", 0)
-
-	streamName := fmt.Sprintf("subscriber-%s", sub.GetName())
-	s.Env().StreamInput(streamName).Publish(sub.GetInput(), attrs)
-}
 
-func (s SubscriberTestSuite) DbTestCase(testCase DbSubscriberTestCase) SubscriberTestCase {
-	modelId, err := mdl.ModelIdFromString(testCase.ModelId)
+	var v SubscriberTestCase
+	var subscriberTestCaseType = reflect.ValueOf(&v).Elem().Type()
```
Comment (ajscholl) [inline pkg/test/suite/testcase_subscriber.go:60]: What is the difference to `reflect.ValueOf(v).Type()`?

File: pkg/cloud/aws/awsv2_middleware_retry.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	return &ErrRetryAttempsExceeded{
+		ResourceName: resourceName,
+		Attempts:     attempts,
+		DurationTook: durationTook,
+		Err:          err,
+	}
+}
+
+func (e *ErrRetryAttempsExceeded) Error() string {
+	return fmt.Sprintf("sent request to resource %s failed after %d retries in %s: %s", e.ResourceName, e.Attempts, e.DurationTook, e.Err)
```
[the remark this answers is not in the document]
Comment (j4k4) [reply pkg/cloud/aws/awsv2_middleware_retry.go:70]: go does it automatically: https://play.golang.org/p/eb0lR1odHsi

File: pkg/cloud/aws/awsv2_middleware_retry.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	stats := getAttemptInfo(ctx)
+
+	if stats == nil {
+		stats = &attemptInfo{
+			start: time.Now(),
+		}
+	}
+
+	stats.count++
+	ctx = smithyMiddleware.WithStackValue(ctx, attemptInfoKey{}, stats)
```
[the remark this answers is not in the document]
Comment (j4k4) [reply pkg/cloud/aws/awsv2_middleware_retry.go:48]: good catch, changed

File: pkg/cloud/aws/awsv2_middleware_retry.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+func (e *ErrRetryMaxElapsedTimeExceeded) Error() string {
+	return fmt.Sprintf("sent request to resource %s failed after %d retries in %s: retry max duration %s exceeded: %s", e.ResourceName, e.Attempts, e.DurationTook, e.DurationMax, e.Err)
+}
+
+func (e *ErrRetryMaxElapsedTimeExceeded) Unwrap() error {
+	return e.Err
+}
+
+func IsErrRequestDurationExceeded(err error) bool {
```
[the remark this answers is not in the document]
Comment (j4k4) [reply pkg/cloud/aws/awsv2_middleware_retry.go:103]: added

File: pkg/cloud/aws/awsv2_middleware_retry.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+			resouceName: resourceName,
+		}
+		ctx = setAttemptInfo(ctx, info)
+
+		if maxElapsedTime > 0 {
+			deadline := clock.Now().Add(maxElapsedTime)
+			ctx, cancel = context.WithDeadline(ctx, deadline)
+
+			defer func() {
+				cancel()
```
[the remark this answers is not in the document]
Comment (j4k4) [reply pkg/cloud/aws/awsv2_middleware_retry.go:130]: indeed

File: pkg/test/env/container_runner.go [hunk 7]
```diff
@@ -100,48 +105,44 @@ func NewContainerRunner(config cfg.Config, logger mon.Logger) (*containerRunner,
...
 
-				containers[skeleton.id()] = container
+					skeleton.containers[name] = container
 
-				return nil
-			}
-		}(skeletons[i]), "can not run container %s", skeletons[i].id())
+					return nil
+				}
+			}(skeletons[i], name, description), "can not run container %s", skeletons[i].id())
```
[the remark this answers is not in the document]
Comment (j4k4) [reply pkg/test/env/container_runner.go:134]: done

File: pkg/test/env/factory_localstack.go [hunk 8]
```diff
@@ -43,6 +44,10 @@ func (f *localstackFactory) Detect(config cfg.Config, manager *ComponentsConfigM
 
 	services := make([]string, 0)
```
[the remark this answers is not in the document]
Comment (j4k4) [reply pkg/test/env/factory_localstack.go:45]: but we want a slice of service names here? there is no check involved. do i get you wrong?

File: pkg/test/suite/run.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+func suiteFindTestCases(t *testing.T, suite TestingSuite) (map[string]testCaseRunner, error) {
+	var err error
+	var testCases = make(map[string]testCaseRunner)
+	var methodFinder = reflect.TypeOf(suite)
+
+	for i := 0; i < methodFinder.NumMethod(); i++ {
+		method := methodFinder.Method(i)
+
+		if ok, _ := regexp.MatchString("^Test", method.Name); !ok {
```
[the remark this answers is not in the document]
Comment (j4k4) [reply pkg/test/suite/run.go]: that's easier indeed

File: pkg/test/suite/run.go [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+	}
+
+	if len(testCases) == 0 {
+		return
+	}
+
+	if suiteOptions.envIsShared {
+		runTestCaseWithSharedEnvironment(t, suite, suiteOptions, testCases)
+	} else {
+		runTestCaseWithIsolatedEnvironment(t, suite, testCases, suiteOptions)
```
[the remark this answers is not in the document]
Comment (j4k4) [reply pkg/test/suite/run.go:42]: changed

File: pkg/test/suite/run.go (hunk 1 above)
Comment (j4k4) [reply pkg/test/suite/run.go:52]: already dropped

File: pkg/test/suite/testcase_subscriber.go (hunk 2 above)
Comment (j4k4) [reply pkg/test/suite/testcase_subscriber.go:60]: that one results in:
```
panic: reflect: call of reflect.Value.Type on zero Value [recovered]
	panic: reflect: call of reflect.Value.Type on zero Value
```

File: pkg/cloud/aws/awsv2.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+)
+
+type ClientHttpSettings struct {
+	Timeout time.Duration `cfg:"timeout" default:"0"`
+}
+
+type ClientRetrySettings struct {
+	InitialInterval time.Duration `cfg:"initial_interval" default:"50ms"`
+	MaxInterval     time.Duration `cfg:"max_interval" default:"10s"`
+	MaxElapsedTime  time.Duration `cfg:"max_elapsed_time" default:"15m"`
```
[the remark this answers is not in the document]
Comment (j4k4) [reply pkg/cloud/aws/awsv2.go:25]: What defaults would you go with?

[the remark this answers is not in the document]
Comment (j4k4) [reply pkg/cloud/aws/awsv2.go:25]: What defaults would you use?

File: pkg/test/suite/testcase_subscriber.go (hunk 2 above)
Comment (ajscholl) [reply pkg/test/suite/testcase_subscriber.go:60]: Ahh, that is an interface, yeah, makes sense then

**Review by ajscholl — APPROVED** [review]

**Pull request merged**